Application boot and testing bootstrap changes

hypeJunction opened the pull request:
My last 5 cents and I really can't see Elgg code anymore.

hypeJunction committed:
chore(boot): allow application to reboot without loosing setups

Adds Includer::requireFileOnce() which requires the file and
stores the returning setup value if it is a Closure.

This allows the tests to replace an instance of an Application
without loosing setup information. This is especially
important with files that should only be required once
(e.g. core files or plugin start.php) because they contain
function declarations as well as hook/event registrations.

hypeJunction committed:
feat(plugins): start.php in core plugins now returns closures

Plugins are advised to return a closure from start.php instead of
placing logic into runtime.

hypeJunction committed:
chore(tests): do not enable plugins during tests

Tests will now run with whatever plugin setup that was active
when tests have started.

Travis builds will now run only with core managed plugins enabled.

hypeJunction committed:
chore(tests): add applicaiton bootstrap test

hypeJunction committed:
chore(tests): simplify testing bootstrap

mrclay commented on engine/classes/ElggPlugin.php:
```diff
@@ -865,7 +865,10 @@ public function start($flags) {
-				require_once "{$this->path}/start.php";
+				$result = Includer::requireFileOnce("{$this->path}/start.php");
+				if ($result instanceof \Closure) {
+					$result(_elgg_services()->hooks->getEvents(), _elgg_services()->hooks);
```
👎  We shouldn't expose our internal event API objects to plugins.

mrclay replied:
https://github.com/hypeJunction/Elgg/pull/1

mrclay committed:
chore(core): keep library setups associated with Application

Don't pass internal API objects to plugin `start.php` files.

hypeJunction commented on engine/tests/classes/Elgg/IntegrationTestCase.php:
```diff
@@ -143,18 +118,26 @@ public static function tearDownAfterClass() {
+		foreach ($this->_testing_hooks as $type => $hooks) {
+			foreach ($hooks as $hook => $handlers) {
+				foreach ($handlers as $handler) {
+					$this->assertTrue($app->_services->hooks->hasHandler($type, $hook, $handler));
```
Does this make sense? I had an afterthought that this might be incorrect

mrclay replied:
Why make assertions in tearDown?

hypeJunction replied:
To make sure tests don't alter the state of hooks. I guess it's redundant now that we start a new application for each test

mrclay replied:
I suppose we could directly compare the output of getAllHandlers() afterwards? I don't know how strict that needs to be.

hypeJunction replied:
I tried that and there are some hooks that use closures with injected session and various timestamps, so it doesn't work

hypeJunction committed:
chore(tests): improve performance of integration tests

Various improvements to speed up integration test suite.

hypeJunction committed:
chore(tests): add Translator tests

hypeJunction commented on engine/classes/Elgg/Application.php:
```diff
@@ -772,7 +772,19 @@ public static function test() {
+		if ($app) {
+			// To speed up database testing, we assign __testing metadata to all created entities
+			// We delete them at shutdown
+			register_shutdown_function(function() {
```
This is hacky but will have to do for now

hypeJunction committed:
chore(tests): upgrade to PHPUnit 5.7

hypeJunction committed:
chore(tests): remove count assertions from subtype tests as they lead to fragile tests

hypeJunction committed:
chore(travis): remove memcached config after build is complete

hypeJunction committed:
chore(seeds): reduce the number of seeded entities

hypeJunction commented:
I don't know what's up with the HHVM build, and why the tests are taking forever to complete.

hypeJunction committed:
chore(tests): reorganize settings in test suites

ELGG_SETTINGS_FILE path in ENV now always take precedence, which
allows running tests with custom configuration.

Shuffles around some settings to speed up travis builds

hypeJunction commented:
Flushing the cache for each test seems to be culprit.

mrclay commented:
#11185 Suggesting we drop HHVM.

hypeJunction commented:
For some reason memcached is enabled on HHVM builds

hypeJunction commented:
@mrclay I've done all I could here. Something is throwing somewhere, but not everywhere, and I don't have what to do with it. Take a look to see if you see something I don't see.

hypeJunction commented:
Perhaps new PHPUnit is stricter about time execution limit https://phpunit.de/manual/current/en/risky-tests.html. I see it to 0 in bootstrap.php, but it doesn't seem to be respected

hypeJunction commented:
Replaced by #11188
